database/migrations/2021_08_09_211715_add_ratings_to_movie_model.php

```diff
@@ -0,0 +1,39 @@
+<?php
+
+use Illuminate\Database\Migrations\Migration;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\Schema;
+use App\Models\Movie;
+
+class AddRatingsToMovieModel extends Migration
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::table('movies', function (Blueprint $table) {
+            $table->decimal('rating', 2, 2, true)->nullable(false)->default(0);
```

Comment:
You can also add an index on `rating` field so sorting will work much faster.

Reply:
Sure. The thing about indexes are that they are a trade-off. Do you have a branch with the index added, so that you can show me a query plan or debug-bar to demonstrate?

Reply:
https://github.com/rm-yakovenko/Laravel-Challenge-Movie-Table-Eloquent/compare/main...rm-yakovenko:pr-36

There are several changes.

1. Fix precision in migration and add an index on `ratings.movie_id`.
2. Instead of command, update movie's rating on review created/updated/deleted. This should keep the cache live without delays.
3. Create 20k movies and 50k reviews. Seeding data can take some time.

Benchmarking
`ab -n 100 <url>`

Method | Time per request (lower is better) | Diff baseline
-- | -- | --
Caching with index | 18.097 | 0.00%
Caching no index | 24.52 | 35.49%
Povilas' solution | 137.473 | 659.65%
Caching with index (2x movies) | 18.59 | 2.72%


Adding new movies almost doesn't change overall time per request. I think that caching is the best option if you're going to have lots of data.

Feel free to ask any questions.

Reply:
I broadly really like it. Thanks for showing figures. To get an explain you can toSql the query and then prepend explain, although it does look much faster and the model observer is a nice touch.

Reply:
@rm-yakovenko I updated this to use your suggestion of an index for the rating field, but missed out several of your other changes.

* seeder alterations
* moving limit / pagination into model
* decimal size 8,2 (I did change to 4,2 as 2,2 did seem to have issues). I Believe that might be better addressed by having DBMS be more permissive to truncate the extra data it receives, but 4,2 seems to avoid any complications.
* model observer (this will mean although we limit reads, writes come out of our control. For that approach I might sooner use raw SQL and be more efficient using a CTE. For movie ratings, I don't think writes would be so often, but it is still an amplification and coupling on movies, which I haven't yet come around to).

I Did alter the controller method to get 110 models, instead of 1010 by playing with your query; although I am not thrilled with the get needing to be passed down to the controller, I really wanted to avoid the need to `->get()` in that controller, but it looks like efficiency has led to coordination. In Ruby on Rails, the get is implicit. I wonder if there is a way to get this using PHP & Laravel. Implicit Get on access of a builder as an iterator?

Reply:
> I broadly really like it. Thanks for showing figures. To get an explain you can toSql the query and then prepend explain, although it does look much faster and the model observer is a nice touch.

```sql
explain select * from `movies` order by `rating` desc limit 100
```

There are 20k movies in my local DB. 

**With index**

id | select_type | table | type | possible_keys | key | key_len | ref | rows | Extra
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
1 | SIMPLE | movies | index |   | movies_rating_index | 4 |   | **100** |  


**Without index**

id | select_type | table | type | possible_keys | key | key_len | ref | rows | Extra
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
1 | SIMPLE | movies | ALL |   |   |   |   | **20112** | Using filesort

With index DB only iterates over 100 rows.

Reply:
> I wonder if there is a way to get this using PHP & Laravel. Implicit Get on access of a builder as an iterator?

I think no. You should call the `get` method explicitly to get an iterator.